### ISSUE TITLE
CFP: External endpoints

### ISSUE THREAD

**Issue by JonasKs** [cilium/cilium#31768]
## Cilium Feature Proposal: External endpoints

**Problem**
We have external resources, such as Cisco ACI EPGs, Azure vNet, Azure ASGs, private VPN tunnels etc which we would like to create CiliumNetworkPolicies with.  
The current solution, merged in #30708 only works with AWS Security Groups. 

**Features**
A way to create custom resolvers, in order to extend `fromGroups`/`toGroups` with any external resource, even private ones.

**Initial thoughts**  
I'm not sure how to make this generic, but we like the way [Crossplane Functions](https://docs.crossplane.io/latest/concepts/composition-functions/#how-composition-functions-work) allow us to extend the Crossplane framework. This would allow us to host our own pod, which is tightly coupled to Cilium, and that pod could resolve the IP addresses.

We also posted a comment in Slack [here](https://cilium.slack.com/archives/C01JALNQAR1/p1712218917450519).

CC: @stianfro, @TerjeLafton, @khguettler

**Comment by Alex-Waring** [comment]
I like the idea but I think it's worth thinking about time to resolve and fallback as well. At the moment the AWS resolver runs on a semi arbitrary timer, and should it be used on scale I am sure we would see AWS API rate limiting.

If you're also building a large network that is making API calls out to an arbitrary endpoint then Cilium policy resolution becomes closely tied to the response of those other features. The scalable way to do this might be to register these as endpoints with a new Cilium resource, that way the resolution of each endpoint can be controlled individually. Not sure what the maintainers think about an additional resource though.

**Comment by JonasKs** [comment]
Good thoughts! Exactly how endpoints is created in k8s is probably not the first problem to tackle, but rather how Cilium would allow for external endpoints.   

As long as there exists a CRD we could create, such as `CiliumExternalEndpoint`, and this resource could be used in `CiliumNetworkPolicies`, our use-case would be solved.   

Example:
```yaml
apiVersion: cilium.io/vx
kind: CiliumExternalEndpoint
metadata:
  name: my-azure-endpoint
  labels:
    k8s:io.cilium.k8s.policy.cluster: default
    system: azure
    type: vnet
spec:
  toCIDRSet:
    - "10.242.0.100/32"
    - "10.242.0.108/32"
    - "10.242.0.105/32"
```

**Comment by JonasKs** [comment]
@soggiest , a little ping from the Microsoft Build conversation 😊

**Comment by JonasKs** [comment]
Hey @soggiest , did you ever get the chance to look at this?

**Comment by JonasKs** [comment]
I've discussed this more on [slack](https://cilium.slack.com/archives/C2B917YHE/p1721029855614309), external endpoints are intended to live in `CiliumCIDRGroup` (CCG), implemented in #24638. 

The only problem with the current implementation is CNP only can reference CCGs directly by name, not resolve them based on selectors with labels. This design don't scale well when we sync in CCGs from external systems, and we want to ensure our CNPs resolve them automatically (e.g. a new CCG of VPN servers for Customer1 is added, we'd like to resolve them with `type: vpn, customer: customer1`.)

I'll close this issue, and carry over the discussion to a more appropriate issue.
